sopel/modules/clock.py

```diff
@@ -55,72 +57,89 @@ def setup(bot):
     bot.config.define_section('clock', TimeSection)
 
 
-@commands('t', 'time')
-@example('.t America/New_York')
+@module.commands('t', 'time')
+@module.example('.t America/New_York')
+@module.example('.t Exirel')
+@module.example('.t #sopel')
 def f_time(bot, trigger):
-    """Returns the current time."""
+    """Returns the current time.
+
+    The command takes an optional parameter: it can be either a timezone, a
+    nick, or a channel.
+    """
     if trigger.group(2):
-        zone = get_timezone(bot.db, bot.config, trigger.group(2).strip(), None, None)
-        if not zone:
-            bot.say('Could not find timezone %s.' % trigger.group(2).strip())
-            return
+        argument = trigger.group(2).strip()
+        channel_or_nick = tools.Identifier(argument)
+
+        if channel_or_nick in bot.users:
+            zone = get_nick_timezone(bot.db, channel_or_nick)
+            if zone is None:
+                bot.say('Could not find timezone for nick %s.' % argument)
+                return
+        elif channel_or_nick in bot.channels:
```

Comment:
I have to say, this approach seems cute at first glance until you realize that it makes getting the time for a user who is offline, or a channel that the bot knows but is not currently joined to, impossible.

While it's easy to decide if the input is a channel name (cast it to `Identifier` and check `is_nick`), the line between nickname and timezone name is less clear. A non-channel input containing `/` is almost certainly a timezone name (standard IRC doesn't allow that character in nicks, but often does allow it in channel names). But things like `CDT` are valid as nicks AND timezone names, so we'd need to prioritize one or the other.

How about this resolution order?

* Input is channel name?
  * Lookup channel TZ
* Not a channel
  * Input contains slash?
    * Attempt to validate and return as zone name
  * No slash
    * Input <= 4 characters?
      * Try as zone name first, then fall back to nick lookup
    * Input > 4 characters
      * Try nick lookup

Maybe there's a less convoluted way to do it. Maybe we should do a strict channel -> nick -> zone resolution order without trying to guess based on length or presence of `/`. But I do think it's important that we _at least_ enable the (anecdotally) common use case of checking what time it is for someone who is offline, e.g. checking whether they're offline because it's the middle of the night for them.

Reply:
I think that we should turn to the [Zen of Python](https://www.python.org/dev/peps/pep-0020/):

> In the face of ambiguity, refuse the temptation to guess.

I'd suggest to create new commands, like `.tn <Nick>`  and ` .tc <Channel>`  instead of trying to be super smart about the input of `.t <zone>`.

* `.t`: timezone from the trigger (nick, then channel, then default)
* ` .t <zone>` : always check for a timezone. If not found, try to see if something else exist, and warn the user that there are specific commands for that
* `.tn`  get trigger's user timezone
* `.tc` get trigger's channel timezone
* `.tn <nick>` and `.tc <channel>`: nick and channel's timezone

It would be simpler, with less guessing and more accurate and predictable behaviours.

Reply:
My gut feeling is that what we have right now is too convenient to give up that easily. If `.t` can continue trying to do the right thing, it's pretty easy to at least get the channel/nick part of things right, and let nicks shadow timezone names if they exist in Sopel's database.

I would be much happier with a single new command, `.tz`, which takes only a timezone name and does no guessing, for use when `.t`'s guessing finds a shadowing nick. (I already know you at least like this idea, from IRC. :P)

Reply:
Let's do this.

Reply:
What do you think now?

Reply:
Hard-coded `.command` bad; use `help_prefix`! (I tell _everyone_ this…)

Once I thought about it a bit, the weird logic to only output an error if the zone is not found _and_ the given nick/channel is in the bot's user/channel tracking makes sense. Looks good!

Reply:
Oh right. I'll use `help_prefix`. Totally forgot about that. :disappointed: